<a href="https://colab.research.google.com/github/shwet369/Cosine-Similarity-Chatbot-main/blob/https%2Fgithub.com%2Fshwet369%2F-_Data-science/Bert_Embedding_Vector_Understabding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install transformers

## Use case

- Understanding the model

- Here we are not using any task

- Here we read the model

- We understand about tokenizer,vocabulary and embeddings

In [2]:
from transformers import BertModel, BertTokenizer
# Every model has its own tokenizer
# Its own words dictinary
# Embeddings

In [3]:
model=BertModel.from_pretrained('bert-base-uncased')
# model name is = 'bert-base-uncased'

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

In [4]:
dir(model)

['T_destination',
 '__annotations__',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_apply',
 '_assisted_decoding',
 '_auto_class',
 '_autoset_attn_implementation',
 '_backward_compatibility_gradient_checkpointing',
 '_backward_hooks',
 '_backward_pre_hooks',
 '_beam_search',
 '_buffers',
 '_call_impl',
 '_check_and_enable_flash_attn_2',
 '_check_and_enable_sdpa',
 '_compiled_call_impl',
 '_constrained_beam_search',
 '_contrastive_search',
 '_convert_head_mask_to_5d',
 '_copy_lm_head_original_to_resized',
 '_create_repo',
 '_dispatch_accelerate_model',
 '_dola_decoding',
 '_expand_inputs_for_generation',

In [5]:
model.framework

'pt'

In [6]:
model.generation_config

In [7]:
model.name_or_path

'bert-base-uncased'

In [8]:
model.num_parameters()

109482240

In [9]:
model.base_model

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(30522, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSdpaSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False

- It will take each and every word

- It will give the word embeddings of each word

- Assume that a sentence has 100 words

- will get 100 word embeddings each vector has 768 values

- 100 words means   (100,768)

- Will apply one pool layer : average of all vectors === activation function

- Again the final sentence word embeddings is a size of 768

In [10]:
# directory embedding
model.embeddings

BertEmbeddings(
  (word_embeddings): Embedding(30522, 768, padding_idx=0)
  (position_embeddings): Embedding(512, 768)
  (token_type_embeddings): Embedding(2, 768)
  (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
  (dropout): Dropout(p=0.1, inplace=False)
)

In [11]:
model.encoder

BertEncoder(
  (layer): ModuleList(
    (0-11): 12 x BertLayer(
      (attention): BertAttention(
        (self): BertSdpaSelfAttention(
          (query): Linear(in_features=768, out_features=768, bias=True)
          (key): Linear(in_features=768, out_features=768, bias=True)
          (value): Linear(in_features=768, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (output): BertSelfOutput(
          (dense): Linear(in_features=768, out_features=768, bias=True)
          (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
      (intermediate): BertIntermediate(
        (dense): Linear(in_features=768, out_features=3072, bias=True)
        (intermediate_act_fn): GELUActivation()
      )
      (output): BertOutput(
        (dense): Linear(in_features=3072, out_features=768, bias=True)
        (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=T

In [12]:
for i in model.named_parameters():
    print(i)

Streaming output truncated to the last 5000 lines.
         3.7748e-03, -9.4684e-03,  5.1092e-03, -1.5236e-03, -7.8516e-03,
         3.0148e-02, -4.6574e-03, -1.5652e-02, -1.8414e-02, -4.8857e-03,
         1.2938e-02,  1.3619e-03,  1.6399e-02,  3.7851e-02,  4.8976e-04,
         2.2249e-03,  3.5374e-02, -4.6371e-02,  3.1519e-02,  1.1191e-02,
         3.5942e-02, -3.4612e-03, -8.5393e-03, -1.8504e-02,  1.7272e-02,
         5.3372e-02, -1.5619e-02, -2.8099e-02,  1.0773e-03,  2.1760e-02,
         3.8248e-02, -4.7646e-03,  1.9283e-02, -6.7653e-03, -1.7880e-02,
         3.9084e-03,  3.5228e-02,  3.6506e-03,  7.3232e-03,  3.7074e-03,
         1.6152e-03,  2.7036e-02, -1.6412e-03, -4.4253e-03, -2.2397e-03,
         1.6211e-02,  4.3293e-03,  2.4323e-03, -2.3251e-02,  2.2354e-02,
         8.7723e-03, -1.9025e-02, -2.3427e-02, -1.9457e-02,  1.3040e-02,
         9.2839e-03,  1.6676e-02, -2.2715e-02, -4.9381e-02, -5.5905e-03,
        -2.1866e-02,  1.8809e-03,  9.7890e-03,  1.1617e-02, -8.3639e-04,


In [13]:
#for i in model.naparameters():
    #print(i)

In [14]:
#for i in model.named_parameters():
    #print(i)

# coding round  sql
# ML : Resume
# Manger round

## Bert Tokenizer

In [15]:
sentence='we are learning bert embeddings'

In [16]:
tokenizer=BertTokenizer.from_pretrained('bert-base-uncased')
tokenizer

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

BertTokenizer(name_or_path='bert-base-uncased', vocab_size=30522, model_max_length=512, is_fast=False, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}

- SEP token is used to seperate the sentences

    - Hello how are you. Im learning BERT

    - Hello how are you [SEP] Im learning BERT

- CLS token is used to close the the sentence

    - Hello how are you [SEP] Im learning BERT [CLS]

- Max length of tokens is 512

- In each sentence the model takes 512 words

- Hello how are you. Im learning BERT

- the number of words :7

- but we required 512, after 7 will add [PAD] tokens

- PAD tokens value is zero


In [17]:
tokenizer.vocab
# all the words and index
# the words consider as keys

OrderedDict([('[PAD]', 0),
             ('[unused0]', 1),
             ('[unused1]', 2),
             ('[unused2]', 3),
             ('[unused3]', 4),
             ('[unused4]', 5),
             ('[unused5]', 6),
             ('[unused6]', 7),
             ('[unused7]', 8),
             ('[unused8]', 9),
             ('[unused9]', 10),
             ('[unused10]', 11),
             ('[unused11]', 12),
             ('[unused12]', 13),
             ('[unused13]', 14),
             ('[unused14]', 15),
             ('[unused15]', 16),
             ('[unused16]', 17),
             ('[unused17]', 18),
             ('[unused18]', 19),
             ('[unused19]', 20),
             ('[unused20]', 21),
             ('[unused21]', 22),
             ('[unused22]', 23),
             ('[unused23]', 24),
             ('[unused24]', 25),
             ('[unused25]', 26),
             ('[unused26]', 27),
             ('[unused27]', 28),
             ('[unused28]', 29),
             ('[unused29]', 30),
  

In [18]:
len(tokenizer.vocab)
# understand the type : dictionary type
# keys are tokens only

30522

In [19]:
tokenizer.vocab.keys()  # only words

# machine
# learning
# machinelearning: machine ##learn ##ing: machine learnig

odict_keys(['[PAD]', '[unused0]', '[unused1]', '[unused2]', '[unused3]', '[unused4]', '[unused5]', '[unused6]', '[unused7]', '[unused8]', '[unused9]', '[unused10]', '[unused11]', '[unused12]', '[unused13]', '[unused14]', '[unused15]', '[unused16]', '[unused17]', '[unused18]', '[unused19]', '[unused20]', '[unused21]', '[unused22]', '[unused23]', '[unused24]', '[unused25]', '[unused26]', '[unused27]', '[unused28]', '[unused29]', '[unused30]', '[unused31]', '[unused32]', '[unused33]', '[unused34]', '[unused35]', '[unused36]', '[unused37]', '[unused38]', '[unused39]', '[unused40]', '[unused41]', '[unused42]', '[unused43]', '[unused44]', '[unused45]', '[unused46]', '[unused47]', '[unused48]', '[unused49]', '[unused50]', '[unused51]', '[unused52]', '[unused53]', '[unused54]', '[unused55]', '[unused56]', '[unused57]', '[unused58]', '[unused59]', '[unused60]', '[unused61]', '[unused62]', '[unused63]', '[unused64]', '[unused65]', '[unused66]', '[unused67]', '[unused68]', '[unused69]', '[unused7

In [20]:
tokenizer.vocab_size

# tokenizer.vocab ==== enterire words with index
# tokenizer.vocab.keys() ==== only words
# tokenizer.vocab_size

30522

In [21]:
tokenizer.vocab['hello'],tokenizer.vocab['##～']

(7592, 30521)

In [22]:
tokenizer.vocab['Hello'] # error
# vocabulary does not have 'Hello'

KeyError: 'Hello'

In [23]:
tokenizer.vocab['machine']

3698

In [24]:
tokenizer.vocab['machine learning'] # two words will not take

KeyError: 'machine learning'

In [25]:
sentence

'we are learning bert embeddings'

In [26]:
tokenizer.tokenize('machine')

['machine']

In [27]:
tokenizer.tokenize('Machine') # M === > m  model name: uncased

['machine']

In [28]:
tokenizer.tokenize('machine learning')

['machine', 'learning']

In [29]:
tokens=tokenizer.tokenize('Machinelearning')
tokens

['machine', '##lea', '##rn', '##ing']

In [30]:
tokenizer.vocab['embeddings']
# embddings is not there

KeyError: 'embeddings'

In [31]:
tokenizer.tokenize('embeddings')

['em', '##bed', '##ding', '##s']

# Sentence tokenize

In [32]:
sentence='we are learning bert embeddings'
tokens=tokenizer.tokenize(sentence)
tokens

['we', 'are', 'learning', 'bert', 'em', '##bed', '##ding', '##s']

In [33]:
tokenizer.vocab['##bed'],tokenizer.vocab['bed']

(8270, 2793)

In [34]:
#tokenizer.vocab['instagram']
tokenizer.tokenize('instagram')

['ins', '##tagram']

In [35]:
#tokens

In [36]:
tokens

['we', 'are', 'learning', 'bert', 'em', '##bed', '##ding', '##s']

In [37]:
# two tokens we need to add to convey that two or more senteces
# cls  and sep
tokens1=['[CLS]']+tokens+['[SEP]']
tokens1

# model will takecare all these things automatically

['[CLS]',
 'we',
 'are',
 'learning',
 'bert',
 'em',
 '##bed',
 '##ding',
 '##s',
 '[SEP]']

In [38]:
#tokens2=['[cls]']+tokens+['[sep]']
#tokens2 # generally we take CLS and SEP

In [39]:
tokens3=['[CLS]']+tokens+['[SEP]']+['[PAD]']+['[PAD]']
tokens3
# l1+l2+l3
# every sentence vector should be in equal length
# we need to pad

['[CLS]',
 'we',
 'are',
 'learning',
 'bert',
 'em',
 '##bed',
 '##ding',
 '##s',
 '[SEP]',
 '[PAD]',
 '[PAD]']

**Attention mask**

- Token is present read as 1

- PAD token read as 0

In [40]:
#attention_mask where ever tokens present =1 for pad tokens zero
# pads are use to match token length only
attention_mask=[1 if i!='[PAD]' else 0 for i in tokens3]
attention_mask

# this is very fundamental concept in bert

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0]

- Fisrt we get the tokens

- Next we added extra tokens like CLS,SEP and PAD

- Then we are creating an attention mask : which has 1 and 0

## get the token ids

- we need to retrive the index of each token based on vocab

In [41]:
dir(tokenizer)

['SPECIAL_TOKENS_ATTRIBUTES',
 '__annotations__',
 '__call__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__len__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_add_tokens',
 '_added_tokens_decoder',
 '_added_tokens_encoder',
 '_additional_special_tokens',
 '_auto_class',
 '_batch_encode_plus',
 '_batch_prepare_for_model',
 '_bos_token',
 '_call_one',
 '_cls_token',
 '_convert_id_to_token',
 '_convert_token_to_id',
 '_convert_token_to_id_with_added_voc',
 '_create_repo',
 '_decode',
 '_decode_use_source_tokenizer',
 '_encode_plus',
 '_eos_token',
 '_eventual_warn_about_too_long_sequence',
 '_eventually_correct_t5_max_length',
 '_from_pretrained',
 '_get_files_timestamps',
 '_get_padding_truncation_strate

In [42]:
tokens3

['[CLS]',
 'we',
 'are',
 'learning',
 'bert',
 'em',
 '##bed',
 '##ding',
 '##s',
 '[SEP]',
 '[PAD]',
 '[PAD]']

In [43]:
tokenizer.vocab['we']

2057

In [44]:
token_ids=tokenizer.convert_tokens_to_ids(tokens3)
token_ids

[101, 2057, 2024, 4083, 14324, 7861, 8270, 4667, 2015, 102, 0, 0]

when sentence is equal to 512 padding is required or not sir..

sentence words >512

words=512

words<512

Truncate=True pad=True

In [45]:
tokenizer.convert_ids_to_tokens(token_ids)

['[CLS]',
 'we',
 'are',
 'learning',
 'bert',
 'em',
 '##bed',
 '##ding',
 '##s',
 '[SEP]',
 '[PAD]',
 '[PAD]']

In [46]:
tokenizer.convert_tokens_to_string(tokens3)

'[CLS] we are learning bert embeddings [SEP] [PAD] [PAD]'

- If sentence is available divide into tokens

- Tokens to token ids

- Suppose we have token ids then we can get tokens

- Suppose we have tokesn then we can get senetence

In [47]:
#tokenizer.convert_ids_to_tokens(token_ids)

In [48]:
tokenizer.vocab['we'],tokenizer.vocab['are'],tokenizer.vocab['learning']

(2057, 2024, 4083)

we have total 30522 tokens we have, which is our vocabulary

for each token we have one vector having 768 values

- Firstly we have sentence

- From sentence will get tokens(words)

- From tokens will get ids

- From ids we get the mask

- In order to get embeddings , we required token_ids and attention mask

In [49]:
import torch
torch.tensor(token_ids) # 1D
torch.tensor(token_ids).unsqueeze(0) # 2D

tensor([[  101,  2057,  2024,  4083, 14324,  7861,  8270,  4667,  2015,   102,
             0,     0]])

In [50]:
import torch
ids=torch.tensor(token_ids).unsqueeze(0)
mask=torch.tensor(attention_mask).unsqueeze(0)

In [51]:
import numpy as np
l=[1,2,3]
np.array(l)

array([1, 2, 3])

In [52]:
ids

tensor([[  101,  2057,  2024,  4083, 14324,  7861,  8270,  4667,  2015,   102,
             0,     0]])

In [53]:
mask

tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0]])

In [54]:
len(tokens3)

12

In [55]:
model

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(30522, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSdpaSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False

In [56]:
out=model(ids,attention_mask=mask)
#dir(out)
out

BaseModelOutputWithPoolingAndCrossAttentions(last_hidden_state=tensor([[[ 0.0092,  0.1259, -0.0047,  ..., -0.4423,  0.1906,  0.6233],
         [ 0.3883, -0.0681, -0.3392,  ..., -0.1198,  1.0829, -0.4945],
         [ 0.1727,  0.1222, -0.2289,  ..., -0.2507,  0.2388,  0.4446],
         ...,
         [ 0.5717,  0.2227, -0.3944,  ...,  0.1266, -0.7046, -0.2170],
         [-0.2574, -0.2615,  0.0443,  ...,  0.0920, -0.0838, -0.0943],
         [-0.2018, -0.2175,  0.0417,  ...,  0.0432, -0.1286, -0.0609]]],
       grad_fn=<NativeLayerNormBackward0>), pooler_output=tensor([[-0.9170, -0.3431, -0.5738,  0.7184,  0.2774, -0.0901,  0.8430,  0.1892,
         -0.2136, -1.0000, -0.0374,  0.6389,  0.9881,  0.1148,  0.9344, -0.5937,
          0.1084, -0.4943,  0.2394, -0.5427,  0.6472,  0.9996,  0.5199,  0.2971,
          0.3666,  0.8450, -0.6475,  0.9381,  0.9606,  0.7622, -0.6632,  0.1369,
         -0.9893, -0.1750, -0.7070, -0.9922,  0.3434, -0.7639,  0.1226,  0.0986,
         -0.9186,  0.2575,  0.99

In [57]:
out[0].shape

torch.Size([1, 12, 768])

In [58]:
#out.items()
out.last_hidden_state
#out[0].shape

# we have 12 tokens are there
# this gives each token embeddvector
# vector has 768 , weights
# 1 batch
# 12 tokens
# 768 values
# 1,12,768

tensor([[[ 0.0092,  0.1259, -0.0047,  ..., -0.4423,  0.1906,  0.6233],
         [ 0.3883, -0.0681, -0.3392,  ..., -0.1198,  1.0829, -0.4945],
         [ 0.1727,  0.1222, -0.2289,  ..., -0.2507,  0.2388,  0.4446],
         ...,
         [ 0.5717,  0.2227, -0.3944,  ...,  0.1266, -0.7046, -0.2170],
         [-0.2574, -0.2615,  0.0443,  ...,  0.0920, -0.0838, -0.0943],
         [-0.2018, -0.2175,  0.0417,  ...,  0.0432, -0.1286, -0.0609]]],
       grad_fn=<NativeLayerNormBackward0>)

In [59]:
out.last_hidden_state.shape
# out[0]

torch.Size([1, 12, 768])

In [60]:
ids

tensor([[  101,  2057,  2024,  4083, 14324,  7861,  8270,  4667,  2015,   102,
             0,     0]])

In [61]:
out.pooler_output.shape
#out[1].shape

torch.Size([1, 768])

In [62]:
out[0] # same as out.last_hidden_state
out[1] # same as out.pooler_output

tensor([[-0.9170, -0.3431, -0.5738,  0.7184,  0.2774, -0.0901,  0.8430,  0.1892,
         -0.2136, -1.0000, -0.0374,  0.6389,  0.9881,  0.1148,  0.9344, -0.5937,
          0.1084, -0.4943,  0.2394, -0.5427,  0.6472,  0.9996,  0.5199,  0.2971,
          0.3666,  0.8450, -0.6475,  0.9381,  0.9606,  0.7622, -0.6632,  0.1369,
         -0.9893, -0.1750, -0.7070, -0.9922,  0.3434, -0.7639,  0.1226,  0.0986,
         -0.9186,  0.2575,  0.9999, -0.5148,  0.3140, -0.2090, -1.0000,  0.1931,
         -0.9005,  0.4248,  0.3755,  0.3924,  0.0979,  0.4084,  0.4009, -0.0047,
         -0.2242,  0.1171, -0.2254, -0.5230, -0.5921,  0.2967, -0.4044, -0.9114,
          0.2170,  0.0368,  0.0386, -0.2435, -0.1261, -0.1515,  0.8507,  0.2650,
          0.0603, -0.8321,  0.1317,  0.1692, -0.5126,  1.0000, -0.3662, -0.9794,
          0.5056,  0.1933,  0.3892,  0.3814,  0.0023, -1.0000,  0.4273, -0.0590,
         -0.9905,  0.0956,  0.4547, -0.0980,  0.2895,  0.5202, -0.3766, -0.3789,
         -0.2373, -0.3929, -

In [63]:
out[1][0][:10]

tensor([-0.9170, -0.3431, -0.5738,  0.7184,  0.2774, -0.0901,  0.8430,  0.1892,
        -0.2136, -1.0000], grad_fn=<SliceBackward0>)



**setence_output** : Is simply an array (representation) of the last layer hidden representation of each token which will be of size (batch_size, seq_len, hidden_size)

**pooler_output:** Is representation/embedding of CLS token passed through some more layers Bertpooler, linear/dense and activation function. It is recommended to use this pooler_output as it contains contextualize information of whole sequence.

In [64]:
import torch
from transformers import BertTokenizer, BertForMaskedLM

# Load pre-trained model and tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')
model = BertForMaskedLM.from_pretrained('bert-base-uncased')

# Example text with a masked token
text = "The cat sat on the [MASK]."

# Tokenize input text
tokenized_text = tokenizer.tokenize(text)
print("Tokenized text:", tokenized_text)

# Convert token to vocabulary indices
masked_index = tokenized_text.index('[MASK]')
indexed_tokens = tokenizer.convert_tokens_to_ids(tokenized_text)

# Convert input to PyTorch tensors
tokens_tensor = torch.tensor([indexed_tokens])

# Predict all tokens
with torch.no_grad():
    outputs = model(tokens_tensor)
    predictions = outputs[0]

# Get the predicted token
predicted_index = torch.argmax(predictions[0, masked_index]).item()
predicted_token = tokenizer.convert_ids_to_tokens([predicted_index])[0]
print("Predicted token:", predicted_token)


BertForMaskedLM has generative capabilities, as `prepare_inputs_for_generation` is explicitly overwritten. However, it doesn't directly inherit from `GenerationMixin`. From 👉v4.50👈 onwards, `PreTrainedModel` will NOT inherit from `GenerationMixin`, and this model will lose the ability to call `generate` and other related functions.
  - If you're using `trust_remote_code=True`, you can get rid of this warning by loading the model with an auto class. See https://huggingface.co/docs/transformers/en/model_doc/auto#auto-classes
  - If you are the owner of the model architecture code, please modify your model class such that it inherits from `GenerationMixin` (after `PreTrainedModel`, otherwise you'll get an exception).
  - If you are not the owner of the model architecture class, please contact the model code owner to update it.
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'c

Tokenized text: ['the', 'cat', 'sat', 'on', 'the', '[MASK]', '.']
Predicted token: .
